File: 9_maximum_slice_problem/max_double_slice_sum.py
def get_prefix_sum(A):
  store = 0
  ret = []

  for i in A:
    store += i
    ret.append(store)

  return ret

def solution(A):
  A = A[1:-1]
  P = get_prefix_sum(A)
  
  start = 0
  m_diff = A[0]
  stored_start = A[0]
  M_sum = 0

  n = len(P)

  for i in xrange(1, n):
    m_diff = min(m_diff, A[i])
    stored_start = min(stored_start, start + m_diff)
    M_sum = max(M_sum, P[i] - stored_start)
    if P[i-1] < start:
      start = P[i-1]
      m_diff = A[i]

  return M_sum 

File: 09_maximum_slice_problem/max_double_slice_sum.cpp
#include <vector>
#include <algorithm>

using namespace std;

vector<int> prefix_sum(vector<int> &a) {
  int n = a.size();
  vector<int> ret(n + 1, 0);
  for (int i = 0; i < n; i++) {
    ret[i + 1] = ret[i] + a[i];
  }
  return ret;
}

int solution(vector<int> &a) {
  int ret = 0;
  int n = a.size();
  vector<int> psum = prefix_sum(a);
  int min_sum = psum[1];
  int least = a[1];
  int saved = min_sum + least;
  for (int i = 2; i < n-1; i++) {
    if (min_sum > psum[i]) {
      min_sum = psum[i];
      least = a[i];
    }
    least = min(least, a[i]);
    saved = min(saved, min_sum + least);
    ret = max(ret, psum[i + 1] - saved);
  }
  return ret;
}
